Add documentation for IAM keys and buckets (#14)
vshn/provider-exoscale: docs/modules/ROOT/pages/explanations/dev/iamkey-reconciliation.adoc

= Development: IAMKey Reconciliation

This page gives a brief overview how `IAMKeys` are managed in the provider.

Note that these diagrams are on a high-level and don't visualize the path for every possible situation.

== Creating IAMKeys

image::iamkey-create.drawio.svg[]

- In exoscale.com API, display names are not unique, there can be multiple IAM keys with different IAM key IDs that share the same display name.

> - During the first reconciliation we do not check if an iam key with the desired display name exists, since we don't know the iam key ID.

Reviewer: ```suggestion
- During the first reconciliation we do not check if an IAM key with the desired display name exists, since we don't know the iam key ID.
```

- After the first reconciliation, the iam key ID, as generated by the exoscale.com API, is stored in the status.
- In exoscale.com API, IAM keys do not depend on buckets, therefore there are no implications once one of its bucket is deleted.

== Updating IAMKeys

image::iamkey-update.drawio.svg[]

> - Iam key is immutable therefore it cannot be updated.

Reviewer: ```suggestion
- IAM keys in exoscale.com are immutable therefore it cannot be updated.
```

== Deleting IAMKeys

image::iamkey-delete.drawio.svg[]

> - exoscale.com API does not prevent deleting an iam keys if there are buckets still attached to it.
> - If the iam key cannot be deleted in exoscale.com API, the resource remains in a "deletion pending" state.

Reviewer: ```suggestion
- exoscale.com API does not prevent deleting an IAM keys if there are buckets still attached to it.
- If the IAM key cannot be deleted in exoscale.com API, the resource remains in a "deletion pending" state.
```
